test: run patch test for each major release and then to PR

ankush opened the pull request:
Change:

- In the patch test, instead of migrating to the current SHA directly. Migrate to the each major releases one at a time (checking out code at the time of release) and then finally migrate to last commit in PR. 
- Why? When directly migrating, the previous patches will have access to latest code and patches won't run as they'd run on IRL installations, hence missing failures. This change brings it a little bit closer to reality. 

Example migration: 

V10 database -> V12 latest release -> v13 latest release -> ... -> last commit in PR. 

Only the first version is hard coded to v12, this should continue to work for all future major releases as long as they follow tagging convention. 

PS: will update the ERPNext test once this is accepted. 

Depends on https://github.com/frappe/frappe/pull/14293 (with a tagged release)

ankush committed:
test: run patch test to last release and PR

In the patch test, instead of migrating to the current SHA directly.
Migrate to the latest release (with code at the time of release) and
then migrate to last commit in PR.

Why? When directly migrating, the previous patches that would've run
with old code have access to the new code and it's not realistic. This
change brings it a little bit closer to reality. Hence missing failures
that would occur with IRL installations.

gavindsouza commented:
I was just thinking about something similar 😅 . I think we should have multiple stages of the patch build...Migrate the backup from each supported major release, from latest v11 to v12 to latest v13 to develop. How does this sound?

gavindsouza commented on .github/workflows/patch-mariadb-tests.yml:
```diff
@@ -102,4 +102,13 @@ jobs:
           cd ~/frappe-bench/
           wget https://frappeframework.com/files/v10-frappe.sql.gz
           bench --site test_site --force restore ~/frappe-bench/v10-frappe.sql.gz
+          cd apps/frappe && git remote set-url upstream https://github.com/frappe/frappe.git && git fetch --all --tags
+          git tag --sort version:refname \
+            | grep -v "beta" \
+            | tail -1 \
+            | xargs -L 1 git checkout -f
```
This might check out to the latest v12 or v13 releases randomly?

ankush replied:
Nope, it shouldn't. `version` sort in git considers "semantic" sorting and not date based, so the last released tag will be picked up (ignoring the ones with `beta` in tag name) 

You can run it locally if you have all the tags.  `git tag --sort version:refname`


ref: https://git-scm.com/docs/git-tag#Documentation/git-tag.txt---sortltkeygt

ankush commented:
> I was just thinking about something similar 😅 . I think we should have multiple stages of the patch build...Migrate the backup from each supported major release, from latest v11 to v12 to latest v13 to develop. How does this sound?


I did this in this PR and then reverted it to this simpler version because it will fail on previously released bad patches (which can't be fixed anymore 🥲 )

We can do it from next release. Old version -> latest v13 -> next version -> PRs.

gavindsouza commented:
> I did this in this PR and then reverted it to this simpler version because it will fail on previously released bad patches (which can't be fixed anymore 🥲 )

On which version? Does it mean it's not possible to migrate from that older version anymore?

ankush commented:
> On which version? Does it mean it's not possible to migrate from that older version anymore?

Some old patch was failing while migrating to v11.x But there's no way to fix that now. I did it for all the tags, I need to try latest tags of each major version. That would be like v10 db -> v11.9 -> v12.25 -> v13.11 -> PR, sounds good?

gavindsouza commented:
> That would be like v10 db -> v11.9 -> v12.25 -> v13.11 -> PR, sounds good?

Yeah, that's cool-er 👍🏼

ankush committed:
test: migration tests each version at a time

ankush committed:
fix: install future and minor changes

minor changes:
1. quitely checkout new tags
2. avoid changing directory
3. use pip instead of bench to install deps.

surajshetty3416 commented:
Ignored flaky UI tests... those got fixed with https://github.com/frappe/frappe/pull/14316
